client/src/featureform/serving.py

```diff
@@ -126,22 +137,88 @@ def _create_channel(self, host, insecure, cert_path):
 
     def features(self, features, entities):
+        return FeatureServer(self._stub, self._resource_client, features, entities)
+
+
+class FeatureServer:
+    def __init__(self, stub, resource_client, features, entities):
+        self._stub = stub
+        self._resource_client = resource_client
+        self._features = features
+        self._entities = entities
+        self._feature_list = []
+        if self._all_ready():
+            self._get_features()
+    
+    def __getitem__(self, item):
+        self._feature_list[item]
+
+    def _get_features(self):
         req = serving_pb2.FeatureServeRequest()
-        for name, value in entities.items():
+        for name, value in self._entities.items():
             entity_proto = req.entities.add()
             entity_proto.name = name
             entity_proto.value = value
-        for (name, variation) in features:
+        for (name, variation) in self._features:
             feature_id = req.features.add()
             feature_id.name = name
             feature_id.version = variation
         resp = self._stub.FeatureServe(req)
-        return [parse_proto_value(val) for val in resp.values]
+        self._feature_list = [parse_proto_value(val) for val in resp.values]
+    
+    
+    def wait(self, timeout=None):
+        """Waits for all features to be ready before retrieving them.
 
+        **Examples**:
+        ``` py
+            client = ff.ServingClient(local=True)
+            fpf = client.features([("avg_transactions", "quickstart")], {"CustomerID": "C1410926"}).wait(timeout=10)
+            # Run features through model
+        ```
+        Args:
+            timeout int: the maximum number of seconds to wait before throwing an error
 
+        Returns:
+            self (FeatureServer object): an object that acts as a list containing every feature retrieved
+        """
+        for (name, variant) in self._features:
+            self._wait_for_feature(name, variant, timeout)
+        self._get_features()
+        return self
+
+    def _wait_for_feature(self, name, variant, timeout):
+        feature = self._resource_client.get_feature(name, variant, verbose=False)
+        status = feature.get_status()
+        timeout_duration = timedelta(seconds=0)
+        if timeout is not None:
+            timeout_duration = timedelta(seconds=timeout)
+        time_waited = timedelta(seconds = 0)
+        time_started = datetime.now()
+        while (status != "FAILED" and status != "READY") and (timeout is None or time_waited < timeout_duration):
+            feature = self._resource_client.get_feature(name, variant, verbose=False)
+            status = feature.get_status()
+            time.sleep(1)
+            time_waited = datetime.now() - time_started
+        if status == "FAILED":
+            raise ValueError(f'Resource {name}:{variant} status set to failed while waiting')
```

Comment:
while waiting feels unnecessary here. Also would be nice to have the error message of why it failed

Reply:
bump